gapis/api/vulkan/frame_loop.go

```diff
@@ -1182,10 +1238,67 @@ func (f *frameLoop) resetResources(ctx context.Context, stateBuilder *stateBuild
 	return nil
 }
 
+func (f *frameLoop) resetDeviceMemory(ctx context.Context, stateBuilder *stateBuilder) error {
+	for mem := range f.memoryToFree {
+		log.D(ctx, "Free memory %v which was allocated during loop.", mem)
+		memObj := GetState(f.loopEndState).DeviceMemories().Get(mem)
+		if memObj == NilDeviceMemoryObjectʳ {
+			return fmt.Errorf("device memory %s doesn't exist in the loop ending state", mem)
+		}
+		if memObj.MappedLocation().Address() != 0 {
+			stateBuilder.write(stateBuilder.cb.VkUnmapMemory(memObj.Device(), mem))
+		}
+
+		stateBuilder.write(stateBuilder.cb.VkFreeMemory(
+			memObj.Device(),
+			memObj.VulkanHandle(),
+			memory.Nullptr,
+		))
+	}
+
+	for mem := range f.memoryToAllocate {
+		log.D(ctx, "Allcate memory %v which was freed during loop.", mem)
+		memObj := GetState(f.loopStartState).DeviceMemories().Get(mem)
+		if memObj == NilDeviceMemoryObjectʳ {
+			return fmt.Errorf("device memory %s doesn't exist in the loop starting state", mem)
+		}
+
+		stateBuilder.createDeviceMemory(memObj, false)
+		if memObj.MappedLocation().Address() != 0 {
+			stateBuilder.write(stateBuilder.cb.Custom(func(ctx context.Context, s *api.GlobalState, b *builder.Builder) error {
+				addr := memObj.MappedLocation().Address()
+				originalTarget, ok := f.mappedAddress[addr]
+				if !ok {
+					return fmt.Errorf("did not find the original mapped address: %v", addr)
+				}
+
+				newTarget, err := b.GetMappedTarget(value.ObservedPointer(addr))
+				if err != nil {
+					return err
+				}
+				b.Load(protocol.Type_AbsolutePointer, newTarget)
+				b.Store(originalTarget)
+
+				return nil
+			}))
+		}
+	}
```

Comment:
What about binding? When we reallocate device memory that was freed in the last iteration of the loop, we need to make sure that it is rebound to all of the buffers and images that were using it when the loop starts?

Reply:
the bindings for affected images/imageviews/buffer/bufferviews will need to addressed. Those need to be done in another PR.